### PR TITLE
L1C product support

### PR DESCRIPTION
**Pull request by toshaklg**
Not sure if weights should be configurable as well or we can keep them hard-coded.

### CODE REVIEW

File: cm_predict.py [hunk 1]
```diff
@@ -81,15 +81,21 @@ def config_from_dict(self, d, product_name):
...
         if product_name:
             self.product_name = product_name
         else:
             self.product_name = d["product_name"]
-        self.weights = d["weights"]
+        if d["level_product"] == "L2A":
+            self.weights = "l2a__030-0.07.hdf5"
+            self.features = ["AOT", "B01", "B02", "B03", "B04", "B05", "B06", "B08", "B8A", "B09", "B11", "B12", "WVP"]
+        elif d["level_product"] == "L1C":
+            self.weights = "l1c__091-0.42.hdf5"
```
Comment (indrek-sunter) [inline cm_predict.py:93]: I think these should not be hard-coded. Instead, these should be taken from the configuration file.

File: cm_predict.py [hunk 2]
```diff
@@ -200,27 +207,20 @@ def mosaic(self):
...
-        tile_paths = []
 
         for subfolder in os.listdir(self.prediction_product_path):
             if os.path.isdir(os.path.join(self.prediction_product_path, subfolder)):
                 image_list.append(pathlib.Path(os.path.join(self.prediction_product_path, subfolder, "prediction.png")))
-        #image_list = [y for x in file_names for y in pathlib.Path(self.predict_folder).glob(f'**/{x}*.png')]
 
-        image_list.sort(key=lambda var: get_img_entry_id(var))
+        # Sort images by asc
+        image_list.sort(key = lambda var: get_img_entry_id(var))
```
Comment (indrek-sunter) [inline cm_predict.py]: I think usually for keyword arguments, whitespaces are not added around the = symbol.

File: cm_predict.py [hunk 3]
```diff
@@ -236,28 +236,32 @@ def mosaic(self):
...
+        # Get name and index from product name
         date_name = self.product_name.rsplit('_', 4)[0].rsplit('_', 1)[1]
         index_name = self.product_name.rsplit('_', 1)[0].rsplit('_', 1)[-1]
 
-        #Define the output names
-        png_name = self.big_image_product +"/" +"L2A_"+index_name+"_"+date_name +'_KZ_10m.png'
-        tif_name = self.big_image_product +"/" +"L2A_"+index_name+"_"+date_name +'_KZ_10m.tif'
+        # Define the output names
+        png_name = big_image_product + "/" + self.product + "_" + index_name + "_" + date_name + '_KZ_10m.png'
+        tif_name = big_image_product + "/" + self.product + "_" + index_name + "_" + date_name + '_KZ_10m.tif'
```
Comment (indrek-sunter) [inline cm_predict.py]: Perhaps better to use os.path.join() here instead of + "/" + ?

File: cm_predict.py [hunk 4]
```diff
@@ -269,53 +273,50 @@ def mosaic(self):
         png_mos = Image.open(png_name)
         tif_mos = Image.open(tif_name)
 
-
-        #Flip final mosaic horisontally
+        # Flip final mosaic horisontally
```
Comment (indrek-sunter) [inline cm_predict.py]: horizontally*

File: cm_predict.py [hunk 5]
```diff
@@ -269,53 +273,50 @@ def mosaic(self):
...
 
-
             profile.update({"count": 1})
 
             with rasterio.open(tif_name, 'w', **profile) as dst:
-               dst.write(band1, 1)
-
+                dst.write(band1, 1)
 
+        # Adding a version tag
```
Comment (indrek-sunter) [inline cm_predict.py]: Add a version tag*

File: cm_predict.py [hunk 6]
```diff
@@ -269,53 +273,50 @@ def mosaic(self):
...
         tif_img = Image.open(tif_name)
-        tif_img.tag[305] = "CM_PREDICT V. {}; CM_VSM V. {}".format(__version__, self.cm_vsm_version)
+        tif_img.tag[305] = "CM_PREDICT V. {}; CM_VSM V. {}".format(__version__, str(self.cm_vsm_version).strip())
```
Comment (indrek-sunter) [inline cm_predict.py]: I don't think "V. " is needed

File: cm_predict.py [hunk 7]
```diff
@@ -269,53 +273,50 @@ def mosaic(self):
...
         tif_img.save(tif_name,tiffinfo=tif_img.tag)
 
         png_img = PngImageFile(png_name)
         metadata = PngInfo()
-        metadata.add_text("Software", "CM_PREDICT V. {}; CM_VSM V. {}".format(__version__, self.cm_vsm_version))
+        metadata.add_text("Software", "CM_PREDICT V. {}; CM_VSM V. {}".format(__version__, str(self.cm_vsm_version).strip()))
```
Comment (indrek-sunter) [inline cm_predict.py]: Same here, I don't think "V. " is needed

File: data_generator.py [hunk 8]
```diff
@@ -10,14 +10,20 @@
...
-                 png_form=False):
+    def __init__(self, list_indices, path_input, batch_size, features, tile_size, num_classes, product_level,
+                 shuffle=True, png_form=False):
         """ Initialization """
         self.path = path_input
-        self.stds = [0.00085, 0.04, 0.037, 0.035, 0.034, 0.035, 0.033, 0.035, 0.034, 0.054, 0.025, 0.021, 0.0083]
-        self.means = [0.0009, 0.02, 0.02, 0.02, 0.02, 0.041, 0.047, 0.045, 0.06, 0.03, 0.03, 0.022, 0.015]
-        self.min_v = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
-        self.max_v = [0.0039, 0.298, 0.282, 0.266, 0.255, 0.252, 0.246, 0.241, 0.241, 0.249, 0.219, 0.221, 0.076]
+        if product_level == "L2A":
```
Comment (indrek-sunter) [inline data_generator.py:17]: These should also be read from a file, instead of being hard-coded here.

**Review by indrek-sunter — COMMENTED** [review]
I don't like the hard-coded weights filenames, feature lists, and statistics. I think these should be read from the configuration file, from some other metadata file, or inferred from the existing files. For example, since the weights files have l1c or l2a prefixes, we could pick them based on the prefix in filename. Are the epoch numbers and metrics in the filenames needed, or could we have model versions there? Can we assume that there is only one L2A and only one L1C model in the models directory at a time? Or perhaps the code could take the first L2A or L1C model weights from the directory by default (if the weights have not been specified in the configuration file)?

File: cm_predict.py (hunk 1 above)
Comment (toshaklg) [reply cm_predict.py:93]: Keeping bands in the configuration file makes it bulky and not so easy to use, keeping them in a somewhat hidden/separate configuration file might be a solution, but as far as I know they do not change at all.

For picking weights file we need further discussion with @anitera

File: cm_predict.py (hunk 5 above)
Comment (toshaklg) [reply cm_predict.py]: Hmmm, okay

File: data_generator.py (hunk 8 above)
Comment (toshaklg) [reply data_generator.py:17]: Same as bands, we need to discuss it

**Review by anitera — COMMENTED** [review]
Looks fine for me

**Pull request merged**